File: src/com/android/deskclock/settings/AlarmVolumePreference.java
/*
 * Copyright (C) 2015 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.deskclock.settings;

import android.app.Activity;
import android.app.NotificationManager;
import android.content.Context;
import android.database.ContentObserver;
import android.media.AudioManager;
import android.os.Build;
import android.provider.Settings;

import androidx.preference.PreferenceViewHolder;
import androidx.preference.SeekBarPreference;

import android.util.AttributeSet;
import android.util.Log;
import android.view.View;
import android.widget.SeekBar;

import com.android.deskclock.R;
import com.android.deskclock.RingtonePreviewKlaxon;
import com.android.deskclock.data.DataModel;

import static android.content.Context.AUDIO_SERVICE;
import static android.content.Context.NOTIFICATION_SERVICE;
import static android.media.AudioManager.STREAM_ALARM;

public class AlarmVolumePreference extends SeekBarPreference {

    private static final long ALARM_PREVIEW_DURATION_MS = 2000;

    private boolean mPreviewPlaying;
    private ContentObserver mVolumeObserver;
    private final AudioManager mAudioManager;

    public AlarmVolumePreference(Context context, AttributeSet attrs) {
        super(context, attrs);
        mAudioManager = (AudioManager) context.getSystemService(AUDIO_SERVICE);
    }

    @Override
    public void onAttached() {
        super.onAttached();

        setMax(mAudioManager.getStreamMaxVolume(STREAM_ALARM));
        setValue(mAudioManager.getStreamVolume(STREAM_ALARM));
    }

    @Override
    public void onBindViewHolder(PreferenceViewHolder holder) {
        super.onBindViewHolder(holder);
        holder.itemView.setClickable(false);

        Context context = getContext();
        SeekBar localSeekBar = (SeekBar) holder.findViewById(R.id.seekbar);

        mVolumeObserver =
                new ContentObserver(localSeekBar.getHandler()) {
                    @Override
                    public void onChange(boolean selfChange) {
                        // Volume was changed elsewhere, update our slider.
                        localSeekBar.setProgress(mAudioManager.getStreamVolume(STREAM_ALARM));
                    }
                };

        context.getContentResolver()
                .registerContentObserver(Settings.System.CONTENT_URI, true, mVolumeObserver);

        localSeekBar.setOnSeekBarChangeListener(
                new SeekBar.OnSeekBarChangeListener() {
                    @Override
                    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                        if (fromUser) {
                            mAudioManager.setStreamVolume(STREAM_ALARM, progress, 0);
                        }
                        setValue(progress);
                        onSeekbarChanged();
                    }

                    @Override
                    public void onStartTrackingTouch(SeekBar seekBar) {}

                    @Override
                    public void onStopTrackingTouch(SeekBar seekBar) {
                        if (!mPreviewPlaying && seekBar.getProgress() != 0) {
                            // If we are not currently playing and progress is set tonon-zero,
                            // start.
                            RingtonePreviewKlaxon.start(
                                    context, DataModel.getDataModel().getDefaultAlarmRingtoneUri());
                            mPreviewPlaying = true;
                            seekBar.postDelayed(
                                    () -> {
                                        RingtonePreviewKlaxon.stop(context);
                                        mPreviewPlaying = false;
                                    },
                                    ALARM_PREVIEW_DURATION_MS);
                        }
                    }
                });
    }

    @Override
    public void onDetached() {
        super.onDetached();
        getContext().getContentResolver().unregisterContentObserver(mVolumeObserver);
    }

    private void onSeekbarChanged() {
        setEnabled(doesDoNotDisturbAllowAlarmPlayback());
        setIcon(getValue() == 0 ? R.drawable.ic_alarm_off : R.drawable.ic_alarm_small);
    }

    private boolean doesDoNotDisturbAllowAlarmPlayback() {
        final NotificationManager notificationManager =
                (NotificationManager) getContext().getSystemService(NOTIFICATION_SERVICE);
        return notificationManager.getCurrentInterruptionFilter()
                != NotificationManager.INTERRUPTION_FILTER_NONE;
    }
}

File: src/com/android/deskclock/settings/AlarmVolumePreference.kt
/*
 * Copyright (C) 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.deskclock.settings

import android.app.NotificationManager
import android.content.Context
import android.content.Context.AUDIO_SERVICE
import android.content.Context.NOTIFICATION_SERVICE
import android.database.ContentObserver
import android.media.AudioManager
import android.media.AudioManager.STREAM_ALARM
import android.provider.Settings
import android.util.AttributeSet
import android.widget.SeekBar
import androidx.preference.PreferenceViewHolder
import androidx.preference.SeekBarPreference

import com.android.deskclock.R
import android.widget.SeekBar.OnSeekBarChangeListener
import com.android.deskclock.RingtonePreviewKlaxon.start
import com.android.deskclock.RingtonePreviewKlaxon.stop
import com.android.deskclock.data.DataModel.Companion.dataModel


class AlarmVolumePreference(context: Context?, attrs: AttributeSet?) : SeekBarPreference(context, attrs) {
    private var mPreviewPlaying = false

    private var mVolumeObserver: ContentObserver? = null
    private var mAudioManager: AudioManager = context?.getSystemService(AUDIO_SERVICE) as AudioManager

    override fun onAttached() {
        super.onAttached()
        max = mAudioManager.getStreamMaxVolume(STREAM_ALARM)
        value = mAudioManager.getStreamVolume(STREAM_ALARM)
    }

    override fun onBindViewHolder(holder: PreferenceViewHolder) {
        super.onBindViewHolder(holder)
        holder.itemView.isClickable = false

        val context = context
        val localSeekBar = holder.findViewById(R.id.seekbar) as SeekBar

        mVolumeObserver = object : ContentObserver(localSeekBar.handler) {
            override fun onChange(selfChange: Boolean) {
                // Volume was changed elsewhere, update our slider.
                localSeekBar.progress = mAudioManager.getStreamVolume(STREAM_ALARM)
            }
        }

        context.contentResolver
                .registerContentObserver(Settings.System.CONTENT_URI, true, mVolumeObserver as ContentObserver)

        localSeekBar.setOnSeekBarChangeListener(
                object : OnSeekBarChangeListener {
                    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                        if (fromUser) {
                            mAudioManager.setStreamVolume(STREAM_ALARM, progress, 0)
                        }
                        value = progress
                        onSeekbarChanged()
                    }

                    override fun onStartTrackingTouch(seekBar: SeekBar) {}
                    override fun onStopTrackingTouch(seekBar: SeekBar) {
                        if (!mPreviewPlaying && seekBar.progress != 0) {
                            // If we are not currently playing and progress is set tonon-zero,
                            // start.
                            start(
                                    context, dataModel.defaultAlarmRingtoneUri)
                            mPreviewPlaying = true
                            seekBar.postDelayed(
                                    {
                                        stop(context)
                                        mPreviewPlaying = false
                                    },
                                    ALARM_PREVIEW_DURATION_MS)
                        }
                    }
                })
    }

    override fun onDetached() {
        super.onDetached()
        context.contentResolver.unregisterContentObserver(mVolumeObserver!!)
    }

    private fun onSeekbarChanged() {
        isEnabled = doesDoNotDisturbAllowAlarmPlayback()
        this.setIcon(if (value == 0) R.drawable.ic_alarm_off else R.drawable.ic_alarm_small)
    }

    private fun doesDoNotDisturbAllowAlarmPlayback(): Boolean {
        val notificationManager = context.getSystemService(NOTIFICATION_SERVICE) as NotificationManager
        return (notificationManager.currentInterruptionFilter
                != NotificationManager.INTERRUPTION_FILTER_NONE)
    }

    companion object {
        private const val ALARM_PREVIEW_DURATION_MS: Long = 2000
    }
}